### PR TITLE
Add Jsonnet rule engine for DB import

### PR DESCRIPTION
**Pull request by beckerhe**
This is adding a Jsonnet-based rule engine for structured data transformation.

We will use that to read benchmark results from GCS files (JSON or CSV or anything else) and transform this into tabular data for a database import.

### CODE REVIEW

File: tools/db_import/tests/test_rules.py
Comment (pzread) [inline tools/db_import/tests/test_rules.py:1]: We usually put tests together with code, e.g., `db_import/rules.py`, `db_import/rules_test.py`

File: tools/db_import/db_import/rules.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import csv, datetime, io, json, re, os
+from typing import Any, Optional, Callable
+import rjsonnet
+from google.cloud import storage
+from db_import.utils import first_no_except
```
Comment (pzread) [inline tools/db_import/db_import/rules.py:11]: Could you sort imports (the first part, e.g., `from *`, `import *`) in lexicographical order?

File: tools/db_import/db_import/rules.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
+## Copyright 2023 The OpenXLA Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+import csv, datetime, io, json, re, os
```
Comment (pzread) [inline tools/db_import/db_import/rules.py:7]: The style guide suggests using separate lines to import (https://google.github.io/styleguide/pyguide.html#313-imports-formatting)

File: tools/db_import/db_import/rules.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+
+# Applies the given `rule`` to `file`.
+# - Returns False if `rule[filepath_regex]` does not match the filepath of `file`.
+# - If it matches, it evaluates the JSONNet expression `rule[result]` and returns its result (parsed as JSON).
+def apply_rule_to_file(
+    rule: dict,
+    file: storage.Blob,
+    config: dict,
+    snippets: dict[str, str],
```
Comment (pzread) [inline tools/db_import/db_import/rules.py:25]: We might still want to use `Dict[str, str]` as `dict[str, str]` is only supported from Python 3.9. Currently we try to support Python 3.8 (default in Ubuntu 20.04)

File: tools/db_import/db_import/rules.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    presence_check: Optional[Callable[[Any, Any], bool]],
+    dump_files_to: Optional[str],
```
Comment (pzread) [inline tools/db_import/db_import/rules.py:27]: Can we use `pathlib.Path` for this parameter? So we can have its convenient methods below (https://docs.python.org/3/library/pathlib.html#correspondence-to-tools-in-the-os-module)

File: tools/db_import/db_import/utils.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+from typing import TypeVar, Callable, Iterable
+
+T = TypeVar("T")
+R = TypeVar("R")
+
+
+# Returns `function(item)` for the first `item` from `iterable` that didn't raise an exception.
+# If all items raise an exception the exception raised by `function(last_item)` is raised.
+# If the iterable yields no items, `None` is returned.
```
Comment (pzread) [inline tools/db_import/db_import/utils.py:15]: These can be the doc strings of the function

File: tools/db_import/db_import/__init__.py
Comment (pzread) [inline tools/db_import/db_import/__init__.py:1]: For the location of these files, I'm planning to rename the current `build_tools/` to `devtools/`. I'm thinking that might be a good place to put all CI related tools.

File: tools/db_import/db_import/__init__.py [hunk 6]
```diff
@@ -0,0 +1,5 @@
+## Copyright 2023 The OpenXLA Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
```
Comment (pzread) [inline tools/db_import/db_import/__init__.py:5]: I guess we don't need to populate the license in the empty `__init__.py`

**Comment by beckerhe** [comment]
@pzread Thank you for the review. I think I addressed all your comments. PTAL! :slightly_smiling_face:

File: devtools/db_import/db_import/rules.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+def apply_rule_to_file(
+    rule: Dict,
+    file: storage.Blob,
+    config: Dict,
+    snippets: Dict[str, str],
+    presence_check: Optional[Callable[[Any, Any], bool]],
+    dump_files_to: Optional[pathlib.Path],
+):
+  """Applies the given `rule`` to `file`.
```
Comment (pzread) [inline devtools/db_import/db_import/rules.py:33]: nit:
```suggestion
  """Applies the given `rule` to `file`.
```

**Review by pzread — APPROVED** [review]

File: devtools/db_import/db_import/rules_test.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+# Blobs can be registered with `registerBlob` and will then appear as present in the bucket.
+class Bucket:
+
+  def __init__(self):
+    self.contents: dict[str, str] = {}
+
+  def blob(self, name: str) -> Blob:
+    return Blob(name, self, self.contents[name])
+
+  def registerBlob(self, name: str, contents: str):
```
Comment (pzread) [inline devtools/db_import/db_import/rules_test.py:40]: nit:
```suggestion
  def register_blob(self, name: str, contents: str):
```

**Pull request merged**